hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/security/authentication/util/SubjectUtil.java

```
@@ -0,0 +1,302 @@
+   * @return the result of the action
+   * @param <T> the type of the result
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T doAs(Subject subject, PrivilegedAction<T> action) {
+    if (HAS_CALL_AS) {
+      try {
+        return callAs(subject, privilegedActionToCallable(action));
+      } catch (CompletionException ce) {
+        Throwable cause = ce.getCause();
+        if (cause != null) {
+          throw sneakyThrow(cause);
+        } else {
+          // This should never happen, CompletionException thrown by Subject.callAs
+          // should always wrap an exception
+          throw ce;
+        }
+      }
+    } else {
+      try {
+        return (T) DO_AS.invoke(subject, action);
+      } catch (Throwable t) {
+        throw sneakyThrow(t);
+      }
+    }
+  }
+
+  /**
+   * Maps action to a Callable on Java 18 onwards, and delegates to callAs().
+   * Call Subject.doAs directly on older JVM.
+   *
+   * @param subject the subject this action runs as
+   * @param action action  the action to run
+   * @return the result of the action
+   * @param <T> the type of the result
+   * @throws PrivilegedActionException if {@code action.run()} throws an checked exception.
+   *      The cause of the {@code PrivilegedActionException} is set to the exception thrown
+   *      by {@code action.run()}.
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T doAs(
+      Subject subject, PrivilegedExceptionAction<T> action) throws PrivilegedActionException {
+    if (HAS_CALL_AS) {
+      try {
+        return callAs(subject, privilegedExceptionActionToCallable(action));
+      } catch (CompletionException ce) {
+        try {
+          Exception cause = (Exception) ce.getCause();
+          if (cause instanceof RuntimeException) {
+            throw (RuntimeException) cause;
+          } else {
+            throw new PrivilegedActionException(cause);
+          }
+        } catch (ClassCastException castException) {
```

I don't get this.
Why limit this to ClassCastException ?
Do we expect the type to be changed between instanceOf and the case ?

Both the legacy and new JDK implementation (and the docs) explicitly wrap any Exception into a CompletionException.

At this point we're either not dealing with an Exception (i.e. it's an Error or raw Throwable), or the JDK implementation is broken.

We should just catch any Exception here

---

Non-Exception throwables are not wrapped by either implementation, so we should just let them bubble up.

---

@stoty this code snippet comes from your original patch ... but I think it's reasonable.

`Subject.callAs` catches `Exception` and wraps it in a `CompletionException`, but `CompletionException` accepts `Throwable` in ctor, and `CompletionException.getCause` also returns a `Throwable` instead of `Exception`.

The catch of `ClassCastException` is for line 227, which shouldn't happen logically, though.
```
Exception cause = (Exception) ce.getCause();
```

```
    public static <T> T callAs(final Subject subject,
            final Callable<T> action) throws CompletionException {
        Objects.requireNonNull(action);
        try {
            return ScopedValue.where(SCOPED_SUBJECT, subject).call(action::call);
        } catch (Exception e) {
            throw new CompletionException(e);
        }
    }
```

```
    public CompletionException(Throwable cause) {
        super(cause);
    }
```

As the comment said, it should never happen, I'm fine to either keep or remove it.

---

After second thought, we can simplify it to

```
        Throwable cause = ce.getCause();
        if (cause instanceof RuntimeException) {
          throw (RuntimeException) cause;
        } else if (cause instanceof Exception) {
          throw new PrivilegedActionException((Exception) cause);
        } else {
          throw sneakyThrow(cause);
        }
```

---

nit: The first if clause is redundant, sneakyThrow will already handle that case.

This is mostly equivalent to what the JDK compatibility method does, so this is as good as it gets.

---

`cause instanceof RuntimeException` is necessary, otherwise subclass of RuntimeException goes branch `cause instanceof Exception` ...